### PR TITLE
Add support for cross-vpc cloning

### PR DESCRIPTION
**Pull request by omnibs**
Hi Charles!

I've adapted `stack-mitosis` to support cross-vpc cloning.

### Restore instead of create-replica

I had to use `RestoreDBInstanceFromDBSnapshot`, as RDS won't let us create cross-vpc replicas.

`RestoreDBInstanceFromDBSnapshot` gives us less configuration over the final created instance, requiring more changes to go into `ModifyDBInstance`, but also saves us a `PromoteReadReplica` step.

It also drifts from `instances` being the full state of the world. It's a bit less nice than `CreateDBInstanceReadReplica` in this way.

### Implicit switch between create-replica and restore-snapshot

I made `stack-mitosis` take the same params on both same-vpc and cross-vpc scenarios, using `RestoreDBInstanceFromDBSnapshot` only for the cross-vpc case.

I also made it assume we'll want the latest available snapshot for the `source` database. For us, with daily backups, the time since latest backup is an acceptable drift when cloning. For some it might not be. I opted to keep the tool simple.

### Restore is faster

I timed `stack-mitosis` on an empty 10GB PG DB with no replicas using both methods:

- Same VPC (22min27s)
  - `CreateDBInstanceReadReplica`: Completed after 672.65s with status available
  - `PromoteReadReplica`: Completed after 305.95s with status available
  - `ModifyDBInstance`: Completed after 122.51s with status available
  - `ModifyDBInstance`: Completed after 122.94s with status available (mitosis-staging -> mitosis-staging-old)
  - `ModifyDBInstance`: Completed after 123.28s with status available (temp-mitosis-staging -> mitosis-staging)
  - `DeleteDBInstance`: No wait

- Cross-VPC (16min14s)
  - `RestoreDBInstanceFromDBSnapshot`: Completed after 486.88s with status available
  - `ModifyDBInstance`: Completed after 182.41s with status available
  - `ModifyDBInstance`: Completed after 182.90s with status available
  - `ModifyDBInstance`: Completed after 122.17s with status available
  - `DeleteDBInstance`: No wait

Which makes me think there's a benefit to always using `RestoreDBInstanceFromDBSnapshot` for cloning primaries. I'm game making that change if you don't mind the drawbacks I listed before.

### Other stuff

- I bundled a change bumping the timeout to 3h bc RDS has been acting up lately and we're seeing 2h30min clone times for prod sometimes
- Kaocha didn't work on macOS with the recent change in `master`, I had to revert back to `clojure -A:kaocha`. It works OK on our CI, so I didn't change it in the PR. Might be something with my Clojure too, as I got it from nix.
- I don't know what you use for formatting, I used `cljfmt` and it got me some diffs. Hope my changes didn't deviate too much from standard.
- These were my first lines of Clojure :S
- Kebab-case is cool

### CODE REVIEW

File: src/stack_mitosis/cli.clj
Comment (dgtized) [inline src/stack_mitosis/cli.clj]: I'm wondering if it makes more sense to always check for snapshots, or only load snapshots when needed?

File: src/stack_mitosis/interpreter.clj
Comment (dgtized) [inline src/stack_mitosis/interpreter.clj]: I can address this, but probably clearer to use `get-in`.

File: src/stack_mitosis/interpreter.clj
Comment (dgtized) [inline src/stack_mitosis/interpreter.clj:103]: I have TODO or issue somewhere to make the timeout a commandline flag. Doesn't hurt to increase it though without that feature.

File: src/stack_mitosis/lookup.clj
Comment (dgtized) [inline src/stack_mitosis/lookup.clj:86]: note to self, look into extracting a common helper to handle combining attributes and translations for each of the 4 cases now.

File: src/stack_mitosis/planner.clj
Comment (dgtized) [inline src/stack_mitosis/planner.clj]: note to self, might be worth extracting a helper function for checking same-vpc, which would make it easier to add an explicit test.

File: src/stack_mitosis/policy.clj
Comment (dgtized) [inline src/stack_mitosis/policy.clj:46]: note to self, check if sub-permissions for og, pg, subgrp are shared and need a helper to generate.

File: src/stack_mitosis/predict.clj
Comment (dgtized) [inline src/stack_mitosis/predict.clj]: A little confused here, maybe just clojure styling or verb choice? I tend to avoid using the `#()` functions in a thread, but it looks like this might be a case where the `as->` thread would allow swapping the position of the hidden argument to make it clearer? Or it might just be that the conj at the end shouldn't be in the thread and it would be cleaner to create the instance attributes, and then as a second thread-last, merge those attributes into the source and conj it onto instances?

Anyway I will take a closer look later.

**Review by dgtized — COMMENTED** [review]
Thanks @omnibs! I will take a deeper look later, but this looks pretty good on a first look through. 

I am somewhat amused at a need to switch to using snapshots as that's how the old ruby tool worked at first creation, before it was switched to use create from replica. I think that was an optimization at the time, so there may be different times where each approach is better.

I think making it implicit to switch to restore from snapshots as a first pass makes sense for the current architecture. If this is ever extended to support Aurora, it would need some other cloning operations as well. In the long run I think it needs a way to force one or the other (and provide source region for cross-region clones), but this is good for now. With regards to last snapshot, that makes sense, through it would be trivial to add a flag to force a snapshot before starting the replication process.

I need to add more integration tests with this, but one thing that is worth checking is the tool currently supports both MySQL & Postgres. While most of the APIs function the same for both there are occasional differences that don't show up until runtime so it is worth running through the integration tests manually if you haven't verified for MySQL.  [example_environment.clj](https://github.com/dgtized/stack-mitosis/blob/master/src/stack_mitosis/example_environment.clj#L10) has functions to create a test environment, along with example invocations in [interpreter.clj](https://github.com/dgtized/stack-mitosis/blob/master/src/stack_mitosis/interpreter.clj#L143). I need to add a fully automatic integration test that creates and destroys a mysql or postgres tree but that's the current manual test. This feature is expanding the surface area to test though, so it may require creating another environment example that actually crosses VPC boundaries? That may be out of scope of this PR but it does make me a little nervous to add a feature that lacks instructions for recreating the environment to test. That said, the IAM policy generation feature was tested manually as well, and doesn't have API calls to set the policy. I guess for now maybe just need to document a simple example of a cross-vpc copy in that example environment file?

I will take a closer look at the individual commits later. There are some style issues I might correct (left a few comments for myself) and will check to see how cljfmt vs default emacs formatting plays out. 

One last thought on the permissions that I will have mull over though is it looks like this is *always* grabbing snapshots, but I think it has enough info to only grab snapshots if the vpc's differ? That might be better as that would keep the set of permissions minimal for each use case.

Anyway thanks for adding the feature, I will see about integrating it.

**Comment by omnibs** [comment]
Thanks for the comments!

> One last thought on the permissions that I will have mull over though is it looks like this is always grabbing snapshots, but I think it has enough info to only grab snapshots if the vpc's differ? That might be better as that would keep the set of permissions minimal for each use case.

It is not! Good thinking on keeping permissions minimal. I'll change that next week.

> it may require creating another environment example that actually crosses VPC boundaries?

I'll do that, I hadn't spotted the `example_environment.clj` file tbh! I hope it doesn't get too wild. We'll need two VPCs, and creating a VPC might get involved. Perhaps using two default VPCs from two different regions is a shortcut.

---

Another thought that crossed my mind: I haven't touched the readme at all.

File: src/stack_mitosis/predict.clj
Comment (omnibs) [reply src/stack_mitosis/predict.clj]: `as->` is cool! Gave it a try

File: src/stack_mitosis/cli.clj
Comment (omnibs) [reply src/stack_mitosis/cli.clj]: Added a follow-up commit doing this

File: src/stack_mitosis/interpreter.clj
Comment (omnibs) [reply src/stack_mitosis/interpreter.clj]: Did it in a follow-up!

**Comment by omnibs** [comment]
Note to self: I think these are the calls we'll need to make to create a VPC for `example_environment.clj`:

- `ec2 :CreateVPC`
- `ec2 :CreateSubnet`
- `rds :CreateDBSubnetGroup`

File: src/stack_mitosis/planner.clj
Comment (omnibs) [reply src/stack_mitosis/planner.clj]: Extracted and added tests!

File: src/stack_mitosis/predict.clj
Comment (dgtized) [reply src/stack_mitosis/predict.clj]: Awesome, thank you, that's much nicer. The threading macros are really nice, but when the direction flips back and forth they are a bit more difficult to follow, so the `as->` really helps here.

File: src/stack_mitosis/planner.clj
Comment (dgtized) [reply src/stack_mitosis/planner.clj]: Thank you!

File: src/stack_mitosis/interpreter.clj
Comment (dgtized) [reply src/stack_mitosis/interpreter.clj]: Thanks!

File: src/stack_mitosis/cli.clj
Comment (dgtized) [reply src/stack_mitosis/cli.clj]: This looks much better. I might come back later and look at changing the `source-snapshot` parameter for `replace-tree` to an optional keyword argument to match `tags` and `restart`, but this looks pretty good as is.

File: src/stack_mitosis/cli.clj
Comment (dgtized) [inline src/stack_mitosis/cli.clj]: Oh, just noticed this ordering. I *think* the same-vpc lookup check may need to be guarded by the call to `verify-databases-exist`, as that should report an error if source or target are missing prior to comparing vpc-id. That way the user can distinguish between source or target didn't exist vs vpc-a != vpc-b because vpc-b is nil because target doesn't exist. 

I think that just means lifting the let for calculating same-vpc inside of the `when` for the `verify-databases-exist`? It's getting a little busy in this method though, I can look into splitting it up a bit afterwards to clarify. I'm happy to take a stab at it if that would help.

**Comment by omnibs** [comment]
I had to add an option to force using restore-snapshot. We have DBs with different DBSubnetGroupNames, and AWS won't let us a) restore with a different subnet group name nor b) modify the subnet group name if both source and target groups are in the same VPC.

Plus I found a bug in restore when the source has replicas.

**Comment by omnibs** [comment]
Derp, forgot to cherry-pick the commit, it's this last one (https://github.com/dgtized/stack-mitosis/pull/7/commits/bbe13bfd6bf7a15855cf3e3791e0ae7cff88fd6f).  It was supposed to come before https://github.com/dgtized/stack-mitosis/pull/7/commits/e553e2c8543d8f8c7c196bb5fd0d870ee046f3ac.

I could have force-pushed but it'd make it harder for you to spot which commits are new to keep on reviewing, sorry for the confusion.

File: src/stack_mitosis/cli.clj
Comment (omnibs) [reply src/stack_mitosis/cli.clj]: Did it in the last commit. I agree it's getting busy.

We might be able to simplify in a follow-up PR by always using restore snapshot for the primaries, if you're game.

Comment (dgtized) [reply src/stack_mitosis/cli.clj]: Thanks for moving that check up, much better. I'll take a look at simplifying, appreciate that it would be less complicated if it always used snapshots, but I think would prefer if both were an option. It's possible it's always faster to use a snapshot because of changes that AWS did on the backend in the past year or two, but I know at one point replicas were much quicker than snapshots. I'm also leery about doing it two different approaches by default. Increases the chance that one or the other will stop working from disuse.

I think this feature accelerates the need for a fully automated integration test (even if it costs a few dollars to run each time). Hopefully we can get this merged in soon so other features can be added more piece-meal. Let me know if there is something here I can help with.

**Comment by dgtized** [comment]
I think we are reasonably close, but still some uncertainty on test cases setup for VPC + readme/changleog updates? I guess for the snapshot copy there are 2 cases, one is for the case with two VPC's, and the other is just forcing it to use snapshot on two RDS replication groups in one VPC? Maybe the second is good enough for now + some comments in the test setup to remind that a two VPC example is required to test the automatic logic?

I think the merge conditions are that a) the new functionality is working & tested for your use case, and b) that the testing + documentation is in place to automate the details later. And I guess also c) that an end-to-end test on mysql/postgres of the existing functionality is still happy.

Anyway, not trying to rush, but also not clear if you are paused or waiting on review! Let me know when you want me to take another look or if you want me to test anything on my end.

**Comment by omnibs** [comment]
Hey, sorry for the radio silence.

I made a separate cli option for scaffolding the test environment with 1 or 2 VPCs, but I haven't figured out how to get the current region in cognitec's aws thingy, so we don't hard-code `us-west-2`'s availability zones in the code path where we're forced to create 2 subnets in 2 different AZs to be able to create a `DBSubnetGroup`.

> a) the new functionality is working & tested for your use case

It is!

> b) that the testing + documentation is in place to automate the details later

This is these last few commits + fixing the hard-coded AZ, right?

edit: oh, plus adding to the `README.md`

> c) that an end-to-end test on mysql/postgres of the existing functionality is still happy.

This is a manual run-through you intend to do? Or something I can do and say "it's done!"?

We could change the example-environment to create four sets of DB, for all invariants:
- mysql, same vpc
- mysql, different vpc
- postgres , same vpc
- postgres, different vpc

I'm thinking:
- default vpc: `mitosis-mysql-src` `mitosis-pg-src` `mitosis-mysql-same-vpc-target` `mitosis-pg-same-vpc-target` + their source replicas
- separate vpc: `mitosis-mysql-different-vpc-target` `mitosis-pg-different-vpc-target` + their replicas

To make validating this easier.

**Comment by omnibs** [comment]
Thinking some more about `create-environment`... Would you oppose moving it to terraform? It's better at handling the create/destroy flows for interdependent resources like vpc->subnet->subnet-group->multiple dbs, and it would also parallelize creation of the several different DBs for all invariants.

**Comment by dgtized** [comment]
Thank you, this is quite a bit more. No worries on the radio silence, just wanted to make sure I wasn't leaving you stuck or anything. This is getting a little more complicated than I quite expected, but I think we can get something merged and then potentially circle back on some of the other issues.

Couple of things;

 1) With the region information, I *think* that should be in the environment variable `AWS_REGION`, or if not in the credentials file loaded in the [CLI](https://github.com/dgtized/stack-mitosis/blob/master/src/stack_mitosis/cli.clj#L49). I'm sure there is some way to query it directly from the aws-api client though, but can't seem to find an example. I did see [one aws-api example](https://github.com/cognitect-labs/aws-api/blob/master/examples/ec2_examples.clj) which would at least allow querying the correct AZs in the current region?

2) For the self test, it's fine to add is as a commandline flag for now, however I think this is a good use case for using the [cli's -X approach](https://clojure.org/guides/deps_and_cli), [with options](https://clojure.org/reference/deps_and_cli#_executing_a_function), or specifying an alternate [CLI entrypoint](https://github.com/dgtized/stack-mitosis/blob/master/test.sh#L3) (ie stack-mitosis.self-test could be a new entrypoint). Alternatively, just setting it up to run as an optional integration test suite (long term plan, not worth doing in this PR). I can extract it from the command line later if you just want to stick to this approach for now. As a slight aside, sorry to hear you only have partial editor support for clojure, it's a really nice interaction flow to evaluate blocks inline, well worth experimenting with, but also appreciate that setting that up may not be quick. 

3) For the self-test, I'd love to experiment with expanding the evaluate-plan approach for more AWS resources than RDS, but I think it might be a stretch to try and re-use it as is by swapping clients to ec2 for the test environment. There are a number of architectural changes needed to support that and so it feels a bit dirty for now. As a quick critique, calling blocking evaluate-plans to request a vpc-id *works* and is clever. However, if it matched the more monadic/interpreter approach the RDS actions are using, it should generate all the commands it plans to execute and have a placeholder vpc-id to fill in over the course of the run. That would also help make the VPC setup idempotent. Setting up a general purpose `evaluate-plan` for EC2 is not particularly simple and well out of scope for this though. <em>So for short term, using some blocking calls to `invoke!` directly and not modifying `evaluate-plan` might be enough to create the vpcs with minimal error checking, which is probably sufficient for this use case.</em> Both `interpret` and `evaluate-plan` are pretty specialized for RDS actions right now, so let's avoid re-using them for EC2 actions for now.

As some architectural asides, *if* we wanted it to support multiple clients (ie ec2, rds) correctly, I think that involves moving closer to a dependency injection approach like https://github.com/stuartsierra/component, or https://github.com/tolitius/mount, and then changing invoke to take some map of functions that can create multiple clients and selects which client appropriately based on the command it's trying to invoke. That's also well out of the scope for this PR, but just wanted to put that out there so you were aware. 

So, all that said about invoke and friends, if you think it would be quick and easy to use terraform to document the vpc creation I think that is also reasonable for now. Long term it would be interesting if stack-mitosis supported EC2 interactions, and it's definitely a clever re-use, just worried about supporting it in the short term (and not really looking to attempt to replace terraform). I do hope to add a `fork` command at some point for creating new stacks from existing but that's still mostly in RDS specific.

If you are comfortable with that approach, let's just add enough terraform to generate a vpc->subnet->subnet-group, and then add a parameter to put the test instances inside of a given VPC. Agree that terraform can probably create the instances in parallel much easier, but I would like to add parallel actions at some point and testing that on test suite creation seems like good happy medium? Converse thought is that moving setup/teardown of test suite instances into terraform might be a better reflection of real world use, so I guess up to you, just please don't remove the existing clojure example environment creation as I intend to re-use that for future integration test setup.

4) For manual run through, again, I still need to automate this, but I usually just stand up a mysql environment, and then run stack-mitosis from the CLI to verify it moved everything over. Sometimes I make sure there are a parameter or two on the source stack that are different from the target stack to see if they copy over. Then I repeat that process for postgres. So if you want to do that I'm happy, otherwise I can take a stab on postgres/mysql, and it sounds like you already did that for the vpc case so that all sounds good.

5) Yea let's just add mention of this functionality to the README & Changelog and bump the version to 0.6?

Sorry for writing a novel here! Trying to intermingle both architectural reasoning and intended plans, but leave the choice up to you. There a few optionals mentioned above, but I think for summary I think all that is remaining for merge is:

 Either:
   - [ ] Handle REGION/AZ detection
   - [ ] Switch to just using `invoke!` for the vpc creation
   - [ ] (Optional) Move environment creation into alternative self-test/smoke-test CLI entrypoint or -X flag

 OR

   - [x] Switch to terraform for vpc creation (this may involve rolling back some of the last couple commits)

and then
 
 - [ ] Test the following environments end-to-end
   - [ ] Mysql replica
   - [ ] Postgres replica 
   - [ ] Mysql cross vpc snapshot
   - [ ] Postgres cross vpc snapshot
 - [x] Update readme + changelog and version bump to 0.6

**Comment by omnibs** [comment]
Found some time to come back here, busy week!

Pivoted to terraform, and pushed readme and changelog changes.

Wasn't able to test all invariants today. We reached our DB instance quota. I'm waiting on a support ticket to raise it.

ps: i can squash the commits + reverts after I get a 👍

**Comment by omnibs** [comment]
Alright, tested all invariants! Everything is working.

Terraform served 3 purposes here:
- Create the environments
- Diff the environments after running stack-mitosis on them (so we know the end databases have the same params we created them with)
- Destroy the environments

I think this is good for a new round of review, and again, I can squash my revert commits after we get a 👍 for merging =]

File: src/stack_mitosis/cli.clj
Comment (dgtized) [inline src/stack_mitosis/cli.clj:69]: I still think source-snapshot should be an optional keyword argument to replace-tree instead of a mandatory positional argument, but happy to address that after this PR merges.

File: src/stack_mitosis/interpreter.clj
Comment (dgtized) [inline src/stack_mitosis/interpreter.clj:88]: No need to adjust, just a comment on clojure style, when using `->>` threading macros, if there are no arguments other than the threaded value, it's ok to leave off the parentheses. Ie:

    (->> snapshots :DBSnapshots (sort-by :SnapshotCreateTime) last :DBSnapshotIdentifier)

File: src/stack_mitosis/interpreter.clj
Comment (dgtized) [inline src/stack_mitosis/interpreter.clj:146]: It looks like the argument for missing snapshot has changed from `nil` to `:none`? I know these are in comment blocks, but might be better to be consistent. My preference would be to keep it as `nil` and change the logic calculating a snapshot to return nil on nothing found. I'm guessing the `:none` placeholder was put in to help guard against invalid or missing values, but as it currently stands this example will fall through to the snapshot case (as nil != :none), and then fail because it's an invalid snapshot. 

(the :none check appears to be https://github.com/dgtized/stack-mitosis/pull/7/files#diff-6d5b0062041657017d820d3bafeecad3300531e355a3adb93bc26694a9b17dfdR45)

File: src/stack_mitosis/lookup.clj
Comment (dgtized) [inline src/stack_mitosis/lookup.clj:60]: Definitely not worth holding this up, but I am a little nervous about this copy pasting. My guess is there is a more restricted list of keys for snapshot, but this is unfortunately one of the places where it's only testable by running it against Amazon. My assumption is that some of these keys are returning nil and so it's working but the code gives the impression it would copy the value when it's not clear if it actually would.  I'll see what I can do about consolidating this, but might require some careful testing in the future.

File: terraform/README.md
Comment (dgtized) [inline terraform/README.md]: It looks like this is missing output? Maybe just leave the command sequence or a small subset of expected output? Prefer that to the empty space, but up to you for how to handle it.

File: terraform/outputs.tf
Comment (dgtized) [inline terraform/outputs.tf:10]: This is a nice little affordance to show examples!

File: test/stack_mitosis/policy_test.clj
Comment (dgtized) [inline test/stack_mitosis/policy_test.clj]: Looks like a stray `dbg` was left in the output but that method was removed? Had to remove this in order to run it through CI.

**Review by dgtized — CHANGES_REQUESTED** [review]
Mostly left comments on future issues, but I think the only concern is the missing comments around terraform output, the stray dbg, and the `:none` vs `nil` on snapshot id. 

My preference for snapshot id is to use nil as the missing value, but maybe could guard the positive case looking for a string with the right format id if that would help catch some errors? I think my view there can be summarized as we should change to use snapshot ids if they exist (ie not nil and correct format strings), but in all other cases it should fallback to the path that doesn't care what the snapshot id is? Current behavior is we have to give it the magic value of `:none` in order to use the default? I suspect this was an attempt to use a `Maybe` or `Just` style matching, but it feels like now it has more possible results but we are only checking for a `Nothing` case & not distinguishing between `Maybe` or `Just`.

Anyway, very last set of changes. I'm happy to step in with the nil/none if you want assistance, but I don't know what needs to go in the command output docs for terraform. Oh also, no need to squash commits to remove the reverts. I often would rebase, but happy to keep the whole history.

Checklist:

 - [x] Handle :none/nil distinction on snapshot-id
 - [x] Remove `dbg` breaking CI (note I need to see how to get CircleCI to run on other folks branches automatically)
 - [x] Fill out terraform docs on command output or remove?

Other than that, I think it's good to go! Thanks so much for adding this functionality and bearing with me throughout all my reviews!

File: src/stack_mitosis/lookup.clj
Comment (omnibs) [reply src/stack_mitosis/lookup.clj:60]: Oh, I actually checked them against the [RestoreDBInstanceFromDBSnapshot](https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_RestoreDBInstanceFromDBSnapshot.html) API inputs. It's why I created a separate one. I would have reused the `clone-replica-attributes` otherwise.

File: src/stack_mitosis/cli.clj
Comment (omnibs) [reply src/stack_mitosis/cli.clj:69]: I agree. I didn't know what was up with `:or` in the keyword args and it's not an easy thing to google, so I didn't venture.

File: src/stack_mitosis/interpreter.clj
Comment (omnibs) [reply src/stack_mitosis/interpreter.clj:146]: Fixed!

File: terraform/README.md
Comment (omnibs) [reply terraform/README.md]: Oops, pushed a fix!

File: src/stack_mitosis/lookup.clj
Comment (dgtized) [reply src/stack_mitosis/lookup.clj:60]: Oh ok, I guess enough looked copy pasted, I wasn't sure. Happy if it was directly tested!

File: src/stack_mitosis/cli.clj
Comment (dgtized) [reply src/stack_mitosis/cli.clj:69]: Cool I can handle that then. I think for this case it probably doesn't even need an `:or` now that it's `nil` as I believe that is the default if nothing is specified.

**Comment by omnibs** [comment]
Rebased away the reverts!

**Comment by omnibs** [comment]
Thanks for the reviews and guidance, and thank you for writing stack-mitosis =]

File: src/stack_mitosis/cli.clj
Comment (dgtized) [inline src/stack_mitosis/cli.clj:63]: Style issue, and I'll address when switching to keyword args, but `if` without an else clause defaults to return nil on else.

**Review by dgtized — APPROVED** [review]
Thanks for all the excellent work, and bearing through all the review passes, really happy to add this feature!

**Pull request merged**
